lib/runtime/src/lib.rs

```diff
@@ -291,7 +298,7 @@ pub fn instantiate(wasm: &[u8], import_object: &ImportObject) -> error::Result<I
 /// The output of this function can be controlled by the mutually
 /// exclusive `default-backend-llvm`, `default-backend-singlepass`,
 /// and `default-backend-cranelift` feature flags.
-pub fn default_compiler() -> impl Compiler {
+pub fn default_compiler() -> Box<dyn Compiler> {
```

Comment:
What's the reason this is now a dynamic trait object?

Reply:
I thought it might be because the compiler can't be statically known anymore, but that doesn't seem to be the case

Reply:
The main reason for this is because I wanted `runtime` to be compilable even when no backends where set (via the `backend-` feature).

However, this caused a compilation error in Rust, mentioning that it didn't know the size of Compiler (since for Rust this looked like an empty function).
I tried to resolve it by doing `panic!` at the end of the function body, but the compilation error still happened, so I changed into a Box<dyn Compiler>` that Rust liked it.

Reply:
Perhaps there are other ways of doing this, let me know if you have ideas @MarkMcCaskey!

Reply:
From an efficiency perspective, this doesn't matter at all, but there are some usability concerns.

If users can interact with compilers outside of the `Compiler` trait, then `impl Compiler` is strictly better because it's more direct:

```rust
let compiler: LLVMCompiler = default_compiler();
compiler.llvm_compiler_specific_method();
```

Whereas with `Box<dyn Compiler>` the user has to do some kind of downcasting (I'm not sure how to do this without looking at docs, but I believe it's possible and will require an `unwrap`).

Also to avoid API breaking it might be worth it to make it work...

One way to do this is to add a Fail compiler which panics on every method... though that really raises the question, in what scenario will a user be compiling with no backends and be calling `default_compiler`?  The only scenario I can think of right now is if the user is making a mistake: if that's the case, then perhaps we want to just remove the `default_compiler` function if no compiler backends are enabled 🤔